File: lib/repositories/business_repository.dart
import "package:capital_commons/core/logger.dart";
import "package:capital_commons/models/business.dart";
import "package:capital_commons/models/create_business.dart";
import "package:cloud_firestore/cloud_firestore.dart";

class BusinessRepositoryException implements Exception {
  const BusinessRepositoryException([
    this.message = "An unexpected error occurred",
  ]);
  final String message;
}

const _businessesCollectionName = "businesses";

class BusinessRepository {
  final _firestore = FirebaseFirestore.instance;

  Future<void> createBusiness(CreateBusiness business) async {
    Log.trace("Creating business with id ${business.uid}");
    try {
      await _firestore
          .collection(_businessesCollectionName)
          .doc(business.uid)
          .set(business.toJson());
    } on FirebaseException catch (e) {
      Log.error(
        "FirebaseException occurred while creating business ${business.uid}: $e",
      );
      throw const BusinessRepositoryException("A FirebaseException occurred");
    }
  }

  Future<Business?> getBusinessById(String businessId) async {
    Log.trace("Fetching business with id $businessId");
    try {
      final doc = await _firestore
          .collection(_businessesCollectionName)
          .doc(businessId)
          .get();

      if (!doc.exists) {
        Log.trace("Business with id $businessId not found");
        return null;
      }

      return Business.fromJson(doc.data()!);
    } on FirebaseException catch (e) {
      Log.error(
        "FirebaseException occurred while fetching business $businessId: $e",
      );
      throw const BusinessRepositoryException("A FirebaseException occurred");
    }
  }
}

File: functions/repositories/business_repository.py
from firebase_admin.exceptions import FirebaseError
from firebase_admin.firestore import firestore
from pydantic import ValidationError

from core import singletons
from core.logger import logger
from models.business import Business


class BusinessRepositoryError(Exception):
    pass


class BusinessRepository:
    _firestore_client: firestore.Client = singletons.firestore_client
    businesses_collection_name = "businesses"

    def get_business_by_id(
        self, business_id: str, transaction: firestore.Transaction | None = None
    ) -> Business:
        try:
            result = (
                self._firestore_client.collection(self.businesses_collection_name)
                .document(business_id)
                .get(transaction=transaction)
            )
        except FirebaseError as e:
            logger.error(
                "FirebaseError occurred while getting business",
                error=e,
                business_id=business_id,
            )
            raise BusinessRepositoryError(
                "A FirebaseError occurred while getting business"
            ) from e

        assert isinstance(result, firestore.DocumentSnapshot)

        try:
            return Business.model_validate(result.to_dict())
        except ValidationError as e:
            logger.error(
                "Validation error occurred while converting doc to Business",
                error=e,
                business_id=business_id,
            )
            raise BusinessRepositoryError(
                "A ValidationError occurred while converting doc to Business"
            ) from e

    def decrement_shares_available(
        self, business_id: str, num_shares: int, transaction: firestore.Transaction
    ) -> None:
        logger.debug(
            "Decrementing shares available",
            business_id=business_id,
            num_shares=num_shares,
        )
        assert num_shares > 0
        ref = self._firestore_client.collection(
            self.businesses_collection_name
        ).document(business_id)
        try:
            transaction.update(
                ref, {"sharesAvailable": firestore.Increment(-1 * num_shares)}
            )
        except FirebaseError as e:
            logger.error(
                f"FirebaseError occurred while decrementing shares available: {e}",
                business_id=business_id,
                num_shares=num_shares,
            )
            raise BusinessRepositoryError("A FirebaseError occurred") from e


business_repository = BusinessRepository()
